dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/reservedvalue/RandomPatternValueGenerator.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2004-2022, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.reservedvalue;
+
+import static java.lang.Math.max;
+import static java.lang.System.currentTimeMillis;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+/**
+ * Generate random values from a pattern using pseudo-randomness to pick letters
+ * and digits.
+ *
+ * Pattern:
+ *
+ * <pre>
+ *  x = lower case
+ *  X = upper case
+ *  # = digit
+ *  * = digit or lower case or upper case
+ * </pre>
+ *
+ * @author Jan Bernitt
+ */
+@NoArgsConstructor( access = AccessLevel.PRIVATE )
+public class RandomPatternValueGenerator
+{
+    private static final String STAR = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+    private static final AtomicLong NEXT_SEED = new AtomicLong( currentTimeMillis() );
```

Comment:
Just a question as we would need something similar to create repeatable performance tests in locust.
Using the same `NEXT_SEED` will lead to the same list of random generated values?

Reply:
Same seed always produces the same pseudo-random sequence.

The seed here is a function of current time and a counter. That means it is the current time millis unless that has already been used before. then we add to it so the next requested seed is also unique. When the next seed is requested and some time has passed we again use the current time and so forth.

Reply:
If I understand you correctly you would want to have the possibility to set the seed for testing. Which is no problem but that code needs to be added.

Reply:
Nice.
Yes, I will need this in another project but it should be as easy as injecting the seed